Server/src/main/resources/templates/leaderboard.vm

```diff
@@ -16,13 +16,19 @@
                 <tr>
                     <th>Player</th>
                     <th>Completed vaults</th>
+                    <th>Death counts</th>
+                    <th>Total execution time (ms)</th>
+                    <th>Walk distance</th>
                 </tr>
                 </thead>
                 <tbody>
                     #foreach($row in $stats)
                     <tr>
                         <td>$row.getKey().getUsername()</td>
                         <td>$row.getValue().size()</td>
+                        <td>$row.getValue().size()</td>
```

Comment:
Same comment as above - I'm calling `size()` to display the number of elements in the array of completed vaults

Reply:
would $row.getValue be better or just $row? or should I just use a different name for each stat?

Reply:
If i read this correctly value `$stats` is an array containing the top 25 entries where each entry is a key-value pair of type `{ key: User, value: CompletedVault[] }`.

I would suggest keeping the user as the key and modifying the value to be an object containing stats. So the model I am proposing (as pseudo code):
```
interface UserStats {
  CompletedVaults: CompletedVault[];
  ExecutionTime: int;
  Deaths: int;
}
// where
$stats = { key: User, value: UserStats }
```

Then the view template would look like:
```
#foreach($row in $stats)
  <tr>
    <td>$row.getKey().getUsername()</td>
    <td>$row.getValue().CompletedVaults.size()</td>
    <td>$row.getValue().ExecutionTime</td>
    <td>$row.getValue().Deaths</td>
  </tr>
```

This would require you to modify the `getUserStatsHelper().getTopN()` function on the `GameServer.INSTANCE` to retrieve those values and return the correct model.

Reply:
I addded a getTopNAll to the UserStatsHelper to get all the stats as an arraylist<map.entry<User,Integer>> but now I'm getting a cast error I don't know why though
![image](https://user-images.githubusercontent.com/10409032/55062367-406afa00-504c-11e9-8c21-a0eb7784872d.png)

Reply:
I will spin up your fork and take a look.

Reply:
@woosle1234 When i clone your repo the package fails to build with `mvn package`. Any idea why? The tests seem to complete succesfull, but still it reports a test failure in the `Server` package. I am checked out on the `issue-182` branch

```
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] server                                                             [jar]
[INFO] server_root                                                        [pom]
[INFO] plugin-biomassBlob                                                 [jar]
[INFO] plugin-npc                                                         [jar]
[INFO] plugin-misc-hw                                                     [jar]
[INFO] plugin-cubot                                                       [jar]
[INFO] plugin-radioactiveCloud                                            [jar]
[INFO] plugin-construction                                                [jar]
[INFO] 
[INFO] ---------------------< net.simon987.server:server >---------------------
[INFO] Building server 1.4a                                               [1/8]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.2:compile (default-compile) @ server ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ server ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\src\test\resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.6.2:testCompile (default-testCompile) @ server ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 41 source files to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\test-classes
[INFO] 
[INFO] --- maven-surefire-plugin:2.7.2:test (default-test) @ server ---
[INFO] Surefire report directory: C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports
The system cannot find the path specified.

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running net.simon987.server.assembly.DWTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 2 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 3B 
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.134 sec
Running net.simon987.server.assembly.instruction.AddInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.instruction.AndInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.BrkInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetaeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetaInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetbeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetbInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetccInstructionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec
Running net.simon987.server.assembly.instruction.SetcInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SeteInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetgeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetgInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetleInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetlInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnaeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetnaInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.instruction.SetnbeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnbInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.instruction.SetncInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetneInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetngeInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetngInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnleInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnlInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnoInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetnsInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetoInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetsInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.instruction.SetzInstructionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
Running net.simon987.server.assembly.LabelTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 0 bytes (1 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
SEVERE: Unknown mnemonic "999:"@0
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 

Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 2 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 01 
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.005 sec
Running net.simon987.server.assembly.MemoryTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.OperandTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.003 sec
Running net.simon987.server.assembly.RegisterSetTest
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.RegisterSet set
INFO: DEBUG: trying to set unknown reg index : 3
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.002 sec
Running net.simon987.server.assembly.TestComment
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 0 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 

Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 4 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 3B 00 0C 
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembly job started: 1 lines to parse.
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: Assembled 8 bytes (0 errors)
Mar 27, 2019 6:23:46 PM net.simon987.server.assembly.Assembler parse
INFO: 
00 53 00 50 00 3D 00 00 
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.006 sec
Running net.simon987.server.game.objects.ItemsContainerTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.009 sec

Results :

Tests run: 51, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for server_root 1.4a:
[INFO] 
[INFO] server ............................................. FAILURE [  4.792 s]
[INFO] server_root ........................................ SKIPPED
[INFO] plugin-biomassBlob ................................. SKIPPED
[INFO] plugin-npc ......................................... SKIPPED
[INFO] plugin-misc-hw ..................................... SKIPPED
[INFO] plugin-cubot ....................................... SKIPPED
[INFO] plugin-radioactiveCloud ............................ SKIPPED
[INFO] plugin-construction ................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.953 s
[INFO] Finished at: 2019-03-27T18:23:46+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.7.2:test (default-test) on project server: There are test failures.
[ERROR] 
[ERROR] Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.7.2:test (default-test) on project server: There are test failures.

Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoFailureException: There are test failures.

Please refer to C:\Users\kevin\Projects\Much-Assembly-Required-woosle1234\Server\target\surefire-reports for the individual test results.
    at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution (SurefireHelper.java:74)
    at org.apache.maven.plugin.surefire.SurefirePlugin.execute (SurefirePlugin.java:642)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

Reply:
Not sure I'm running on Windows 10 and works fine